tools/istio-iptables/pkg/dependencies/implementation.go

```diff
@@ -51,51 +50,133 @@ var exittypeToString = map[XTablesExittype]string{
 	XTablesResourceProblem:  "xtables resource problem",
 }
 
-// XTablesWriteCmds contains all xtables commands that do write actions (and thus need a lock)
-var XTablesWriteCmds = sets.New(
-	constants.IPTABLES,
-	constants.IP6TABLES,
-	constants.IPTABLESRESTORE,
-	constants.IP6TABLESRESTORE,
-)
-
 // RealDependencies implementation of interface Dependencies, which is used in production
 type RealDependencies struct {
-	IptablesVersion  IptablesVersion
 	NetworkNamespace string
 	CNIMode          bool
 }
 
 const iptablesVersionPattern = `v([0-9]+(\.[0-9]+)+)`
 
 type IptablesVersion struct {
+	DetectedBinary        string
+	DetectedSaveBinary    string
+	DetectedRestoreBinary string
 	// the actual version
-	version *utilversion.Version
+	Version *utilversion.Version
 	// true if legacy mode, false if nf_tables
-	legacy bool
+	Legacy bool
+	// true if we detected that existing rules are present for this variant (legacy, nft, v6)
+	ExistingRules bool
 }
 
-func DetectIptablesVersion(ver string) (IptablesVersion, error) {
-	if ver == "" {
-		var err error
-		verb, err := exec.Command("iptables", "--version").CombinedOutput()
-		if err != nil {
-			return IptablesVersion{}, err
-		}
-		ver = string(verb)
+func (v IptablesVersion) CmdToString(cmd constants.IptablesCmd) string {
+	switch cmd {
+	case constants.IPTables:
+		return v.DetectedBinary
+	case constants.IPTablesSave:
+		return v.DetectedSaveBinary
+	case constants.IPTablesRestore:
+		return v.DetectedRestoreBinary
+	default:
+		return ""
+	}
+}
+
+// IsWriteCmd returns true for all command types that do write actions (and thus need a lock)
+func (v IptablesVersion) IsWriteCmd(cmd constants.IptablesCmd) bool {
+	switch cmd {
+	case constants.IPTables:
+		return true
+	case constants.IPTablesRestore:
+		return true
+	default:
+		return false
+	}
+}
+
+// Constants for iptables commands
+// These should not be used directly/assumed to be present, but should be contextually detected
+const (
+	iptablesBin         = "iptables"
+	iptablesNftBin      = "iptables-nft"
+	iptablesLegacyBin   = "iptables-legacy"
+	ip6tablesBin        = "ip6tables"
+	ip6tablesNftBin     = "ip6tables-nft"
+	ip6tablesLegacyBin  = "ip6tables-legacy"
+	iptablesRestoreBin  = "iptables-restore"
+	ip6tablesRestoreBin = "ip6tables-restore"
+)
+
+// It is not sufficient to check for the presence of one binary or the other in $PATH -
+// we must choose a binary that is
+// 1. Available in our $PATH
+// 2. Matches where rules are actually defined in the netns we're operating in
+// (legacy or nft, with a preference for the latter if both present)
+//
+// This is designed to handle situations where, for instance, the host has nft-defined rules, and our default container
+// binary is `legacy`, or vice-versa - we must match the binaries we have in our $PATH to what rules are actually defined
+// in our current netns context.
+//
+// Q: Why not simply "use the host default binary" at $PATH/iptables?
```

Comment:
An alternate way to select between the two modes at run time is to use the approach explained in the following repo - https://github.com/kubernetes-sigs/iptables-wrappers

Reply:
I'm not completely familiar with Istio code base. Should we update this code as well? (as its forcing iptables to always use legacy) 
https://github.com/istio/istio/blob/2c8dcd45a39964c6270653acd7605427004cae2e/docker/Dockerfile.base#L29

Reply:
> An alternate way to select between the two modes at run time is to use the approach explained in the following repo - https://github.com/kubernetes-sigs/iptables-wrappers

Yes - as mentioned this will not work for us because we do _not_ always have access to the host binaries because we are not always running in the host FS. `kubelet` does, so they are fine doing it this way. Their heuristic for detecting which binary to use is similar, however.


> I'm not completely familiar with Istio code base. Should we update this code as well? (as its forcing iptables to always use legacy)
> 
> https://github.com/istio/istio/blob/2c8dcd45a39964c6270653acd7605427004cae2e/docker/Dockerfile.base#L29

It doesn't matter, as we will prefer `legacy` or `nft` binaries explicitly, so the "container default" binary is moot.

We no longer need to force a container default, and so can remove that line if we wish however - it won't change anything.

Reply:
> Yes - as mentioned in the PR description this will not work for us because we do _not_ always have access to the host binaries because we are not always running in the host FS.

We dont need access to the host binaries inside the pod. We can add the logic to the Dockerfile. As an example, you can see the following file - https://github.com/submariner-io/submariner/blob/3248d61fd6e20a6cb1d877ce75e30ba2a226faa6/package/Dockerfile.submariner-route-agent#L39

However, one requirement is that the pod should have priviledges to run `iptables` command (I guess the pod will have them, as it needs to program some iptables rules). 

>It doesn't matter, as with this PR we will prefer legacy or nft binaries explicitly, so the "container default" binary is moot - we can't rely on the container default being the same as the host default, so the container default isn't relevant to our selection heuristic.

Right, I understand that. If the container defaults points to appropriate binary then we dont have to make any changes in the Istio `go` code. 
Anyways, I don't have any strong opinions on the approach and am fine with the selection logic inside the pod if it's acceptable to others.

Reply:
> > Yes - as mentioned in the PR description this will not work for us because we do _not_ always have access to the host binaries because we are not always running in the host FS.
> 
> We dont need access to the host binaries inside the pod. We can add the logic to the Dockerfile. As an example, you can see the following file - https://github.com/submariner-io/submariner/blob/3248d61fd6e20a6cb1d877ce75e30ba2a226faa6/package/Dockerfile.submariner-route-agent#L39

Since we are jumping into arbitrary pod network namespaces during runtime in some contexts, we cannot select the correct binary except at runtime, dynamically, at that point. So it is not something we can accurately detect in the Dockerfile.

> Right, I understand that. If the container defaults points to appropriate binary then we dont have to make any changes in the Istio `go` code. Anyways, I don't have any strong opinions on the approach and even the approach of selecting the iptables-nft/legacy inside the pod would work.

The container default *can never* point to the appropriate binary in _all cases_ where the wrapper is invoked, is the problem. No matter what we do with it. So we cannot rely on the system default either way.